### PR TITLE
Define `SpreadElement`

### LINKED ISSUE #7: Formalize SpreadElement
Estree currently doesn't define `SpreadElement`, which is used by esprima and acorn in `ArrayPattern`s and `ArrayExpression`s.

[Example](http://felix-kling.de/esprima_ast_explorer/#/ykp8P3Yy1c/1):

```js
// [a, ...b] = x;
{
  "type": "AssignmentExpression",
  "operator": "=",
  "left": {
    "type": "ArrayPattern",
    "elements": [{
      "type": "Identifier",
      "name": "a"
    }, {
      "type": "SpreadElement",
      "argument": {
        "type": "Identifier",
        "name": "b"
      },
    }],
  },
  "right": {
    "type": "Identifier",
    "name": "x"
  }
}
```

Edit:

I'm not suggesting/implying that it has to be `SpreadElement` or that it has to be used in both cases, just we have to properly define it (them) for `ArrayPattern`s and `ArrayExpression`s.

### PR DESCRIPTION
**Pull request by RReverser**
Fixes #7

### CODE REVIEW

**Comment by RReverser** [comment]
@mikesherov @ariya @dherman Can you please :+1: ?

**Comment by mikesherov** [comment]
Do we have to do anything special for use of `speadElement` as part of a destructuring assignment?

**Comment by mikesherov** [comment]
Also, a general trend I've noticed is the lack of prose describing what these nodes represent. We've merged a few PRs without prose, but I'd like us to get away from that.

Also, do we have to specifically call out that the redefinition of CallExpression and ArrayExpression here? Meaning, maybe an informative note mentioning the difference and that this definition overrides the ES5 definition?

**Comment by RReverser** [comment]
In destructuring assignment it's `RestElement`, and for cases like `someFunc(...x)` and `var ar = [1, ...otherArr, 2]` it's `SpreadElement`.

**Comment by RReverser** [comment]
> Also, do we have to specifically call out that the redefinition of CallExpression and ArrayExpression here? Meaning, maybe an informative note mentioning the difference

It's not redefinition - we already use `extend interface ...` in this DSL in every place where we want to add / redefine particular properties with new semantics.

**Comment by RReverser** [comment]
Just another note - it's not something new, this is how things already work in Acorn and Esprima.

**Comment by RReverser** [comment]
As for SpiderMonkey, it seems to use something like following instead:

```js
interface SpreadExpression <: Node {
    expression: Expression;
}
```

(need @dherman to confirm)

File: es6.md [hunk 1]
```diff
@@ -31,6 +31,18 @@ extend interface VariableDeclaration {
 # Expressions
 
 ```js
+interface SpreadElement <: Node {
+    argument: Expression;
```
Comment (Constellation) [inline es6.md]: It also requires `type: "SpreadElement";`, correct?

Comment (RReverser) [inline es6.md]: Yes, thanks, @Constellation!

**Comment by mikesherov** [comment]
I'll defer to @ariya here.

**Comment by RReverser** [comment]
Would be nice to know opinions of other participants, too.

**Comment by dherman** [comment]
That's right re: SpiderMonkey:
```
js> Reflect.parse("f(1,...a,2)", { loc: false }).body[0].expression.arguments[1]
({loc:null, type:"SpreadExpression", expression:{loc:null, type:"Identifier", name:"a"}})
```
I'll try to submit a patch to fix it soon.

**Comment by dherman** [comment]
I agree that `extend interface` is enough to call out backwards-compatible extensions.

I agree with @mikesherov that there should be prose for each element. :+1: from me for this PR otherwise.

**Comment by RReverser** [comment]
Added prose and warning about SM behavior.

**Comment by RReverser** [comment]
@ariya Please give :+1: so I could merge this.

**Comment by ariya** [comment]
:+1:

**Pull request merged**